### PR TITLE
added task code

### PR DESCRIPTION
**Pull request by IMrCrimeI**

### CODE REVIEW

File: app/src/main/java/com/onix/internship/arch/BaseActivity.kt [hunk 1]
```diff
@@ -11,7 +11,7 @@ import com.google.android.material.snackbar.Snackbar
 abstract class BaseActivity<T : ViewDataBinding>(@LayoutRes private val resId: Int) :
     AppCompatActivity() {
 
-    protected lateinit var binding: T
+    private lateinit var binding: T
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/arch/BaseActivity.kt:14]: Agree, but all this staff should be committed before create PR

File: app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import org.koin.androidx.viewmodel.ext.android.viewModel
+
+class ChoiceFragment : BaseFragment<ChoiceFragmentBinding>(R.layout.choice_fragment) {
+
+    override val viewModel: ChoiceViewModel by viewModel()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.firstPlayerButton.setOnClickListener {
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt]: onclick should triggered in binding - take a look in splashfragment navigation example

File: app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            findNavController().navigate(ChoiceFragmentDirections.actionChoiceFragmentToGameFragment(PlayerMove(false).player))
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt]: there is two repeated code with PlayerMove(false).player and PlayerMove(true).player - rewrite with function

File: app/src/main/java/com/onix/internship/ui/game/GameModel.kt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package com.onix.internship.ui.game
+
+data class GameModel(var gameBoard:Array<IntArray> = Array(3){IntArray(3)})
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/GameModel.kt]: replace int with enum

File: app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+import android.util.Log
+import com.onix.internship.arch.BaseViewModel
+
+class GameViewModel: BaseViewModel() {
+//    val gameBoard = Array(3) { IntArray(3) }
+
+    val gameBoard = GameModel().gameBoard
+
+    var winType = arrayOf(-1, -1, -1)
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt]: should be enum

File: app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    private var boardFilled = 0
+    private var player = 1
+
+    fun updateGameBoard(row: Int, col: Int): Boolean {
+        return if (gameBoard[row - 1][col - 1] == 0) {
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt]: rewrite it without -1 and also without if statement

File: app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            false
+        }
+    }
+
+    fun computer() {
+        Log.d("Cell", findEmptyCells().toString())
+        if (findEmptyCells() != listOf<EmptyCell>()) {
+            val emptyCell = findEmptyCells().random()
+            Log.d("Cell", emptyCell.toString())
+            gameBoard[emptyCell.rowIndex][emptyCell.colIndex] = 2
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt]: enum please

File: app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            arrayElement.forEachIndexed { index, element ->
+                if (element == 0) {
+                    emptyCells.add(EmptyCell(arrayIndex, index))
+                }
+            }
+        }
+        return emptyCells
+    }
+
+    fun winnerCheck(): Boolean {
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt:58]: rewrite it in another way, for example with templates of winning matrix

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+                if (game.updateGameBoard(row, col)) {
+                    invalidate()
+                    game.computer()
+                    invalidate()
+                    if (game.winnerCheck()) {
+                        winningLine = true
+                        invalidate()
+                    }
+//                    if (game.player % 2 == 0) {
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: remove all comments from commit

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+import com.onix.internship.R
+import kotlin.math.ceil
+
+class TicTacToeBoard(context: Context, attrs: AttributeSet?) : View(context, attrs) {
+
+    private val size = 0.2f
+    private var winningLine = false
+    private val paint = Paint()
+    private var cellSize = width / 3
+    private val game = GameViewModel()
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: game should be stored inside viewmodel and view just render model on scren and pass user input into viewmodel

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+            return true
+        }
+        return false
+    }
+
+    private fun drawGameBoard(canvas: Canvas) {
+        paint.color = boardColor
+        paint.strokeWidth = 16f
+
+        for (c in 1..2) {
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: why 1..2 hardcoded ? digging count of cells from GameViewModel

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        }
+    }
+
+    private fun drawX(canvas: Canvas, row: Int, col: Int) {
+        paint.color = xColor
+
+
+        canvas.drawLine(
+            (col + 1) * cellSize - cellSize * size,
+            row * cellSize + cellSize * size,
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt:141]: simplify this calculation with method that calculate cell boundary and return it in Rect

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+            cellSize * 3f, cellSize * 3f, paint
+        )
+    }
+
+    private fun drawWinningLine(canvas: Canvas) {
+        val row = game.winType[0]
+        val col = game.winType[1]
+
+        when (game.winType[2]) {
+            1 -> drawHorizontalLine(canvas, row, col)
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: should be enum

File: app/src/main/java/com/onix/internship/ui/main/MainScreen.kt [hunk 14]
```diff
@@ -21,29 +17,6 @@ class MainScreen : BaseActivity<ActivityMainBinding>(R.layout.activity_main) {
...
 
-    override fun onConfigurationChanged(newConfig: Configuration) {
-        super.onConfigurationChanged(newConfig)
-    }
-
-    override fun onNewIntent(intent: Intent?) {
-        super.onNewIntent(intent)
-    }
-
     override fun setObservers() {}
```
Comment (shadwork) [inline app/src/main/java/com/onix/internship/ui/main/MainScreen.kt]: why setObservers() needed here ?

**Review by shadwork — CHANGES_REQUESTED** [review]
очень хорошо!
не нарушай мввм - вью показывает, модель хранит данные, вью модель содержит логику связей
когда пишешь похожий код - думай о замене на метод

File: app/src/main/java/com/onix/internship/arch/BaseActivity.kt (hunk 1 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/arch/BaseActivity.kt:14]: okay

File: app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt (hunk 2 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt]: done

File: app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt (hunk 3 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/choice/ChoiceFragment.kt]: done

File: app/src/main/java/com/onix/internship/ui/game/GameModel.kt (hunk 4 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/game/GameModel.kt]: done

File: app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt (hunk 5 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/game/GameViewModel.kt]: done

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt (hunk 10 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: done

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt (hunk 11 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: done

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt (hunk 12 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt:141]: done

File: app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt (hunk 13 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/game/TicTacToeBoard.kt]: done

File: app/src/main/java/com/onix/internship/ui/main/MainScreen.kt (hunk 14 above)
Comment (IMrCrimeI) [reply app/src/main/java/com/onix/internship/ui/main/MainScreen.kt]: why no?)

**Pull request merged**